File: C_sources/make_blocks.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <ctype.h>
#include <string.h>
#include <math.h>
#include <time.h>
#include "gthreads.h"

#define WIDTH 800
#define HEIGHT 600
#define NR_BLOCKS 7500

typedef struct position_str{
    int i;
    int j;
}position;

int *matrix;
int *local_threads;
position* pos;

char* itoa(int val, int base){
	static char buf[32] = {0};

	int i = 30;

	for(; val && i ; --i, val /= base)
		buf[i] = "0123456789abcdef"[val % base];

	return &buf[i+1];
}

void read_matrix(void){
    FILE* file;
    file = fopen("../input/matrix/matrix.in", "r");
    if(file == NULL){
        printf("Error reading matrix file\n");
        return;
    }

    for (int i = 0; i < HEIGHT; i++){
        for (int j = 0; j < WIDTH; j++){
            int pos = i * WIDTH + j;
            fscanf(file, "%d", &matrix[pos]);
        }
    }
    fclose(file);

}

void shuffle(position *array, int n)
{
    if (n > 1){
        int i;
        for (i = 0; i < n - 1; i++){
          int j = i + rand() / (RAND_MAX / (n - i) + 1);
          position t = array[j];
          array[j] = array[i];
          array[i] = t;
        }
    }
}

void write_block(int posIBlock, int posJBlock, int index){
    FILE *out;

    char filename_out[19] = "out/blocks/block";
    char* nr;
    nr = itoa(index,10);

    int digits = index / 10 + 1;

    strncat(filename_out, nr, digits);
    //printf("%s\n",filename_out);

    out = fopen(filename_out, "w");
    if(out == NULL){
        printf("Error opening block file\n");
        return;
    }

    double polynom = 0;
    for (int i = 0; i < 8; i++){
        for (int j = 0; j < 8; j++){
            int pos = ((i + posIBlock) * WIDTH) + j + posJBlock;

            fprintf(out, "%d ", matrix[pos]);
            if(i == j){
                polynom += pow((double)matrix[pos], (double)i);
            }
        }
        fprintf(out,"\n");
    }
    fprintf(out, "%lf", polynom);
    fclose(out);
}

int main(){
    matrix = (int*) malloc(WIDTH*HEIGHT*sizeof(int));
    read_matrix();
    gthread_init();

    local_threads = (int*) malloc(NR_BLOCKS * sizeof(int));
    pos = (position*) malloc(NR_BLOCKS * sizeof(position));

    int counterPos = 0;
    int counter = 0;

    for(int i = 0; i<HEIGHT; i+=8){
        for(int j = 0; j<WIDTH; j+=8){
            pos[counterPos].i = i;
            pos[counterPos].j = j;
            counterPos++;
        }
    }

    shuffle(pos, NR_BLOCKS);

    struct timeval start,stop;
    gettimeofday(&start, NULL);
    long long startTime = (long long)start.tv_sec * 1000 + (long long)start.tv_usec / 1000;

    for(int i=0; i<NR_BLOCKS; i++){
        local_threads[counter] = gthread_run(write_block,pos[i].i,pos[i].j,i);
        counter++;
    }

    for(int i = 0;i<counter;i++){
        gthread_join(local_threads[i]);
    }

    gettimeofday(&stop, NULL);
    long long endTime = (long long)stop.tv_sec * 1000 + (long long)stop.tv_usec / 1000;
    long long time = endTime - startTime;

    printf("%lld\n", time);
    free(matrix);
    free(pos);
    free(local_threads);
    return 0;
}

File: GO_sources/make_blocks.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"math/rand"
	"os"
	"strconv"
	"sync"
	"time"
)

type position struct {
    i int
    j int
}

const WIDTH = 800;
const HEIGHT = 600;

var matrix [HEIGHT][WIDTH]int;

func write_block(wg *sync.WaitGroup, iBlock int, jBlock int, count int){
	var filename_out = "out/blocks/block" + strconv.Itoa(count)

	out, _ := os.Create(filename_out)

	bufr := bufio.NewWriter(out)
	var polynom float64 = 0;
	for i := 0; i < 8; i++{
		for j := 0; j < 8; j++{
			fmt.Fprintf(bufr, "%d ", matrix[iBlock + i][jBlock + j])
			if(i==j){
				polynom += math.Pow(float64(matrix[iBlock+i][jBlock+j]), float64(i))
			}
		}
		fmt.Fprintf(bufr, "\n")
		bufr.Flush()
	}
	fmt.Fprintf(bufr, "%f", polynom)
	bufr.Flush()

	out.Close()
	wg.Done()
}

func read_matrix(filename string) {
    file, _ := os.Open(filename)

    bufr := bufio.NewReader(file)

	for i := 0; i < HEIGHT; i++ {
		for j := 0; j < WIDTH; j++ {
			fmt.Fscanf(bufr, "%d", &matrix[i][j])
		}
	}
}

func main() {
	read_matrix("../input/matrix/matrix.in");

	var positions [7500]position;
	var counter int = 0;

	for i := 0; i < HEIGHT; i+=8 {
		for j := 0; j < WIDTH; j+=8 {
			positions[counter].i = i;
			positions[counter].j = j;
			counter++;
		}
	}

	rand.Seed(time.Now().UnixNano())
	rand.Shuffle(len(positions), func(i, j int) { positions[i], positions[j] = positions[j], positions[i] })

	wg := new(sync.WaitGroup)

	t := time.Now()
	start := t.UnixMilli()

	for i := 0; i < len(positions); i++{
		wg.Add(1)
		go write_block(wg, positions[i].i, positions[i].j, i)
	}
	wg.Wait()

	t = time.Now()
	stop := t.UnixMilli()

	diff := stop - start;
	fmt.Println(diff);
}


